src/topology/compacts.lean

```diff
@@ -1,105 +1,120 @@
+
 /-- The type of closed subsets of a topological space. -/
-def closeds := {s : set α // is_closed s}
+structure closeds (α : Type*) [topological_space α] :=
+(carrier : set α)
+(closed' : is_closed carrier)
+
+namespace closeds
+variables {α}
 
-/-- The type of closed subsets is inhabited, with default element the empty set. -/
-instance : inhabited (closeds α) := ⟨⟨∅, is_closed_empty ⟩⟩
+instance : set_like (closeds α) α :=
+{ coe := closeds.carrier,
+  coe_injective' := λ s t h, by { cases s, cases t, congr' } }
 
-/-- The compact sets of a topological space. See also `nonempty_compacts`. -/
-def compacts : Type* := { s : set α // is_compact s }
+lemma closed (s : closeds α) : is_closed (s : set α) := s.closed'
 
-/-- The type of non-empty compact subsets of a topological space. The
-non-emptiness will be useful in metric spaces, as we will be able to put
-a distance (and not merely an edistance) on this space. -/
-def nonempty_compacts := {s : set α // s.nonempty ∧ is_compact s}
+@[ext] protected lemma ext {s t : closeds α} (h : s.1 = t.1) : s = t := set_like.ext' h
 
-/-- In an inhabited space, the type of nonempty compact subsets is also inhabited, with
-default element the singleton set containing the default element. -/
-instance nonempty_compacts_inhabited [inhabited α] : inhabited (nonempty_compacts α) :=
-⟨⟨{default}, singleton_nonempty default, is_compact_singleton ⟩⟩
+instance : has_sup (closeds α) := ⟨λ s t, ⟨s ∪ t, s.closed.union t.closed⟩⟩
+instance : has_inf (closeds α) := ⟨λ s t, ⟨s ∩ t, s.closed.inter t.closed⟩⟩
+instance : has_top (closeds α) := ⟨⟨univ, is_closed_univ⟩⟩
+instance : has_bot (closeds α) := ⟨⟨∅, is_closed_empty⟩⟩
 
-/-- The compact sets with nonempty interior of a topological space. See also `compacts` and
-  `nonempty_compacts`. -/
-@[nolint has_inhabited_instance]
-def positive_compacts: Type* := { s : set α // is_compact s ∧ (interior s).nonempty }
+instance : lattice (closeds α) := set_like.coe_injective.lattice _ (λ _ _, rfl) (λ _ _, rfl)
+instance : bounded_order (closeds α) := bounded_order.lift (coe : _ → set α) (λ _ _, id) rfl rfl
 
-/-- In a nonempty compact space, `set.univ` is a member of `positive_compacts`, the compact sets
-with nonempty interior. -/
-def positive_compacts_univ {α : Type*} [topological_space α] [compact_space α] [nonempty α] :
-  positive_compacts α :=
-⟨set.univ, compact_univ, by simp⟩
+/-- The type of closed sets is inhabited, with default element the empty set. -/
+instance : inhabited (closeds α) := ⟨⊥⟩
 
-@[simp] lemma positive_compacts_univ_val (α : Type*) [topological_space α] [compact_space α]
-  [nonempty α] : (positive_compacts_univ : positive_compacts α).val = univ := rfl
+@[simp] lemma coe_sup (s t : closeds α) : (↑(s ⊔ t) : set α) = s ∪ t := rfl
+@[simp] lemma coe_inf (s t : closeds α) : (↑(s ⊓ t) : set α) = s ∩ t := rfl
+@[simp] lemma coe_top : (↑(⊤ : closeds α) : set α) = univ := rfl
+@[simp] lemma coe_bot : (↑(⊥ : closeds α) : set α) = ∅ := rfl
```

Comment:
These lemmas should come above `instance : lattice`, because then you can use them in place of the `rfl`s there.

Reply:
Eh, but it's longer to write `coe_xxx` than `rfl`. The point of those lemmas is to `rw`, not so much to give the proofs.